grpc_opentracing/grpc_interceptor/open_tracing_client_interceptor.py

```diff
@@ -0,0 +1,173 @@
+    try:
+        tracer.inject(span.context, opentracing.Format.HTTP_HEADERS, headers)
+    except (opentracing.UnsupportedFormatException,
+            opentracing.InvalidCarrierException,
+            opentracing.SpanContextCorruptedException) as e:
+        logging.exception('tracer.inject() failed')
+        span.log_kv({'event': 'error', 'error.object': e})
+        return metadata
+    metadata = () if metadata is None else tuple(metadata)
+    return metadata + tuple((k.lower(), v) for (k, v) in six.iteritems(headers))
+
+
+class OpenTracingClientInterceptor(grpc.UnaryUnaryClientInterceptor,
+                                   grpc.UnaryStreamClientInterceptor,
+                                   grpc.StreamUnaryClientInterceptor,
+                                   grpc.StreamStreamClientInterceptor):
+
+    def __init__(self, tracer, log_payloads):
+        self._tracer = tracer
+        self._log_payloads = log_payloads
+
+    def _start_span(self, method):
+        active_span_context = None
+        active_span = scope.get_active_span()
+        if active_span is not None:
+            active_span_context = active_span.context
+        tags = {
+            ot_tags.COMPONENT: 'grpc',
+            ot_tags.SPAN_KIND: ot_tags.SPAN_KIND_RPC_CLIENT
+        }
+        return self._tracer.start_span(
+            operation_name=method, child_of=active_span_context, tags=tags)
+
+    def _intercept_call(
+            self, client_call_details, request_iterator
+    ):
+        metadata = ()
+        if client_call_details.metadata is not None:
+            metadata = client_call_details.metadata
+
+        # Start a span
+        current_span = self._start_span(client_call_details.method)
+
+        metadata = _inject_span_context(self._tracer, current_span,
+                                        metadata)
+        client_call_details = _ClientCallDetails(
+            client_call_details.method,
+            client_call_details.timeout,
+            metadata,
+            client_call_details.credentials)
+
+        if self._log_payloads:
+            request_iterator = grpc_utils.log_or_wrap_request_or_iterator(
+                span=current_span,
+                is_client_stream=True,
+                request_or_iterator=request_iterator
+            )
+
+        return client_call_details, request_iterator, current_span
+
+    def _callback(self, current_span):
+        def callback(future_response):
+            exception = future_response.exception()
+            if exception is not None:
+                exception = str(exception)
+                current_span.set_tag('error', True)
+                error_log = {'event': 'error', 'error.kind': exception}
+                current_span.log_kv(error_log)
+
+            if self._log_payloads:
+                response = future_response.result()
+                current_span.log_kv({'response': response})
+
+            scope.set_active_span(current_span)
+            scope.end_span()
```

Comment:
this doesn't quite work they way it might be expected. I believe you need to pass the original span from before when you create a new one at https://github.com/opentracing-contrib/python-grpc/pull/14/files#diff-4ecafa65ccad14bb230575320502f261R57

The reason for this is that otherwise if there are further gRPC calls issued by the calling code after this point before it explicitly creates a new span each of these calls will appear as though the previous gRPC was it's parent.

e.g.
```
 - start tracer span
         |
         \--- client grpc call 1
                     |--- server grpc function 1
                     |            \--- ... other tracing calls in server
                     \--- client grpc call 2
                                 |--- server grpc function 2
                                 |            \--- ... other tracing calls in server
                                 \--- client grpc call 3
                                              \--- server grpc function 3
                                                           \--- ... other tracing calls in server
```
instead of
```
 - start tracer span
         |
         |--- client grpc call 1
         |            \--- server grpc function 1
         |                          \--- ... other tracing calls in server
         |--- client grpc call 2
         |            \--- server grpc function 2
         |                          \--- ... other tracing calls in server
         \--- client grpc call 3
                       \--- server grpc function 3
                                     \--- ... other tracing calls in server
```

So the function should take two arguments `def _callback(self, current_span, original_span):` and then do `current_span.finish()` followed by `scope.set_active_span(original_span)`. Since you want to finish the current span, but when exiting the grpc interceptor back to the program code the active span should be set back to what it was before the interceptor was called.